File: frontend/src/lib/api/admin-console.ts
import { ApiError, apiFetch, apiUrl, authHeaders } from "@/lib/api/client";
import { parseBrandVoiceOrderResource } from "@/lib/api/billing";

// 管理员后台 adapter（ADMIN-CONSOLE-UI-0001 · FIX1 已按**真实 BE #165** 逐字段对齐）。
// 契约源：backend/app/schemas/admin_console.py + routes/admin_console.py（merge 618d7b94）。
// 路由前缀 /api/v1/admin/console，统一 require_platform_admin → 403 PLATFORM_ADMIN_REQUIRED。
// 分页一律 page/page_size（响应含 page/page_size）；错误 message 为 BE 中文，UI 原样展示。

const BASE = "/api/v1/admin/console";

export type PlanCode = "free" | "basic" | "huading";
export type TenantStatus = "active" | "suspended" | "closed";
export type AdminTaskFamily = "video" | "reverse_prompt" | "ecom_replicate";
export type AdminTaskStatus = "queued" | "running" | "succeeded" | "failed" | "cancelled";
export type AuditAction = "credits_adjust" | "plan_change" | "status_change" | "voice_slot_assign" | "task_retry";
export type AdminBrandVoiceOrderStatus = "awaiting_fulfillment" | "fulfilled" | "rejected";
export type AdminBrandVoiceOrderAction =
  | { action: "fulfill"; provider_voice_id: string }
  | { action: "reject"; rejection_reason: string };

export interface AdminBrandVoiceOrderRead {
  id: string;
  tenant_id: string;
  ordered_by_user_id: string;
  order_type: "create" | "renew";
  requested_name: string;
  source_audio_asset_id: string;
  existing_brand_voice_id: string | null;
  status: AdminBrandVoiceOrderStatus;
  fulfilled_brand_voice_id: string | null;
  fulfilled_provider_voice_id: string | null;
  rejection_reason: string | null;
  fulfilled_at: string | null;
  expires_at: string | null;
  rejected_at: string | null;
  created_at: string;
  updated_at: string;
  billing: import("@/lib/api/types").BillingSummary;
  refund_disposition: "not_applicable" | "source_subscription_released" | "current_subscription_credited" | "pending_next_subscription";
  refund_grant_status: "pending" | "applied" | null;
  refund_applied_at: string | null;
  source_audio_url?: string | null;
}

export interface AdminBrandVoiceOrderPage {
  items: AdminBrandVoiceOrderRead[];
  total: number;
  page: number;
  page_size: number;
}

function invalidBrandVoiceOrderResponse(): never {
  throw new ApiError("品牌音色订单响应不符合契约。", "INVALID_BRAND_VOICE_ORDER_RESPONSE", 502);
}

function parseAdminBrandVoiceOrder(value: unknown, allowSourceAudioUrl = false): AdminBrandVoiceOrderRead {
  return parseBrandVoiceOrderResource(value, { allowSourceAudioUrl }) ?? invalidBrandVoiceOrderResponse();
}

/** 订阅额度快照（BE AdminSubscriptionSnapshot）。 */
export interface AdminSubscriptionSnapshot {
  id: string;
  total: number;
  used: number;
  reserved: number;
  remaining: number;
}

/** 租户列表行（BE AdminTenantItem——owner_email/plan_code/subscription 均可空）。 */
export interface AdminTenantRow {
  tenant_id: string;
  slug: string;
  name: string;
  status: TenantStatus | string;
  created_at: string;
  owner_email: string | null;
  plan_code: PlanCode | string | null;
  subscription: AdminSubscriptionSnapshot | null;
  task_count: number;
}

export interface AdminTenantPage {
  items: AdminTenantRow[];
  total: number;
  page: number;
  page_size: number;
}

export type AdminTenantSortField = "credits_used" | "created_at" | "balance";
export type AdminSortOrder = "asc" | "desc";

export interface AdminTenantListQuery {
  q?: string;
  plan?: PlanCode | "";
  status?: TenantStatus | "";
  sort?: AdminTenantSortField;
  order?: AdminSortOrder;
  page: number;
  page_size: number;
}

/** 用量流水行（BE AdminUsageItem）。 */
export interface AdminUsageRow {
  id: string;
  created_at: string;
  tenant_id: string;
  tenant_slug: string;
  tenant_name: string;
  capability: string;
  provider: string;
  model: string | null;
  quantity: number;
  unit: string;
  credits: number;
  cost_cents: number;
  status: "reserved" | "settled" | "released" | string;
  video_task_id: string | null;
}

export interface AdminUsagePage {
  items: AdminUsageRow[];
  total: number;
  page: number;
  page_size: number;
}

export interface AdminUsageQuery {
  tenant_id?: string;
  from?: string;
  to?: string;
  capability?: string;
  provider?: string;
  status?: string;
  page: number;
  page_size: number;
}

/** 任务监控行（BE AdminTaskItem——retryable 由 BE 判定，前端不自判 status）。 */
export interface AdminTaskRow {
  id: string;
  task_family: AdminTaskFamily;
  tenant_id: string;
  tenant_slug: string;
  tenant_name: string;
  mode: string;
  label: string | null;
  video_mode: string | null;
  status: AdminTaskStatus;
  progress: number | null;
  error_code: string | null;
  error_message: string | null;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
  duration_seconds: number | null;
  retryable: boolean;
}

export interface AdminTaskPage {
  items: AdminTaskRow[];
  total: number;
  page: number;
  page_size: number;
}

/** 音色槽位（BE AdminVoiceSlotItem——平台池与租户专属合一列表，scope 区分）。 */
export interface AdminVoiceSlotItem {
  speaker_id: string;
  scope: "platform" | "tenant";
  sources: string[];
  tenant_id: string | null;
  tenant_slug: string | null;
  tenant_name: string | null;
  occupied: boolean;
  brand_voice_id: string | null;
  brand_voice_name: string | null;
  brand_voice_status: string | null;
}

export interface AdminVoiceSlots {
  items: AdminVoiceSlotItem[];
  total: number;
  remaining: number;
}

/** 审计日志行（BE AdminAuditLogItem——before/after 可为 null）。 */
export interface AdminAuditRow {
  id: string;
  actor_user_id: string;
  actor_email: string | null;
  actor_tenant_id: string;
  action: AuditAction | string;
  target_tenant_id: string | null;
  target_tenant_slug: string | null;
  target_id: string | null;
  before: Record<string, unknown> | null;
  after: Record<string, unknown> | null;
  reason: string | null;
  created_at: string;
}

export interface AdminAuditPage {
  items: AdminAuditRow[];
  total: number;
  page: number;
  page_size: number;
}

/**
 * 重试回执（BE AdminTaskRetryResponse，202）——FIX1 冻结三态，**无 estimate_basis 字段**：
 * charged=false → 不重复扣费；charged=true+is_estimate=false → credits 实扣（固定价）；
 * charged=true+is_estimate=true → credits 为预计（仅 released avatar_talk，按实际成片时长结算）。
 */
export interface AdminRetryReceipt {
  id: string;
  task_family: AdminTaskFamily;
  tenant_id: string;
  status: "queued";
  progress: number;
  charged: boolean;
  credits: number;
  is_estimate: boolean;
}

function qs(params: Record<string, string | number | undefined>): string {
  const sp = new URLSearchParams();
  for (const [key, value] of Object.entries(params)) {
    if (value !== undefined && value !== "") sp.set(key, String(value));
  }
  return sp.toString();
}

export function fetchAdminTenants(query: AdminTenantListQuery): Promise<AdminTenantPage> {
  return apiFetch<AdminTenantPage>(`${BASE}/tenants?${qs({ ...query })}`, { method: "GET" });
}

export interface AdminTenantDetail {
  tenant: AdminTenantRow;
  recent_tasks: AdminTaskRow[];
  recent_usage: AdminUsageRow[];
  voice_slots: AdminVoiceSlotItem[];
}

export function fetchAdminTenantDetail(tenantId: string): Promise<AdminTenantDetail> {
  return apiFetch<AdminTenantDetail>(`${BASE}/tenants/${encodeURIComponent(tenantId)}`, { method: "GET" });
}

/** 余额增减（delta≠0 + reason 1–500 必填）→ {tenant_id, delta, subscription}。下限保护 422 CREDIT_TOTAL_BELOW_COMMITTED（中文原样展示）。 */
export function adjustTenantCredits(
  tenantId: string,
  input: { delta: number; reason: string }
): Promise<{ tenant_id: string; delta: number; subscription: AdminSubscriptionSnapshot }> {
  return apiFetch(`${BASE}/tenants/${encodeURIComponent(tenantId)}/credits`, { method: "POST", body: input });
}

/** 改套餐（PATCH）。平台租户降级 → 422 CANNOT_DOWNGRADE_PLATFORM_TENANT。 */
export function changeTenantPlan(
  tenantId: string,
  planCode: PlanCode,
  reason?: string
): Promise<{ tenant_id: string; plan_code: PlanCode; subscription: AdminSubscriptionSnapshot }> {
  return apiFetch(`${BASE}/tenants/${encodeURIComponent(tenantId)}/plan`, {
    method: "PATCH",
    body: { plan_code: planCode, ...(reason ? { reason } : {}) }
  });
}

/** 启用/停用（PATCH {active:bool}）→ status active/suspended。停用平台租户 → 422 CANNOT_SUSPEND_PLATFORM_TENANT。 */
export function changeTenantStatus(
  tenantId: string,
  active: boolean,
  reason?: string
): Promise<{ tenant_id: string; status: TenantStatus }> {
  return apiFetch(`${BASE}/tenants/${encodeURIComponent(tenantId)}/status`, {
    method: "PATCH",
    body: { active, ...(reason ? { reason } : {}) }
  });
}

export function fetchAdminVoiceSlots(): Promise<AdminVoiceSlots> {
  return apiFetch<AdminVoiceSlots>(`${BASE}/voice-slots`, { method: "GET" });
}

export async function listAdminBrandVoiceOrders(query: {
  status?: AdminBrandVoiceOrderStatus | "";
  page: number;
  page_size: number;
}): Promise<AdminBrandVoiceOrderPage> {
  const page = await apiFetch<unknown>(`${BASE}/brand-voice-orders?${qs(query)}`, { method: "GET" });
  if (
    typeof page !== "object" || page === null || Array.isArray(page) ||
    Object.keys(page).length !== 4 ||
    !["items", "total", "page", "page_size"].every((key) => Object.prototype.hasOwnProperty.call(page, key))
  ) invalidBrandVoiceOrderResponse();
  const candidate = page as Record<string, unknown>;
  if (!Array.isArray(candidate.items) || !Number.isSafeInteger(candidate.total) || !Number.isSafeInteger(candidate.page) || !Number.isSafeInteger(candidate.page_size)) invalidBrandVoiceOrderResponse();
  return {
    items: candidate.items.map((item) => parseAdminBrandVoiceOrder(item)),
    total: candidate.total as number,
    page: candidate.page as number,
    page_size: candidate.page_size as number
  };
}

export async function getAdminBrandVoiceOrder(orderId: string): Promise<AdminBrandVoiceOrderRead> {
  return parseAdminBrandVoiceOrder(
    await apiFetch<unknown>(`${BASE}/brand-voice-orders/${encodeURIComponent(orderId)}`, { method: "GET" }),
    true
  );
}

export async function resolveAdminBrandVoiceOrder(
  orderId: string,
  action: AdminBrandVoiceOrderAction
): Promise<AdminBrandVoiceOrderRead> {
  const value = await apiFetch<unknown>(`${BASE}/brand-voice-orders/${encodeURIComponent(orderId)}/resolve`, {
    method: "POST",
    body: action
  });
  return parseAdminBrandVoiceOrder(value);
}

/** doubao speaker_id 前端预校验（BE pattern 权威）。 */
export const SPEAKER_ID_PATTERN = /^S_[A-Za-z0-9_-]{1,157}$/;

/** BE AdminTaskStatus 用 succeeded（非 done）——映射到既有 StatusBadge 视觉档（tasks 页 / 租户详情共用）。 */
export const TASK_BADGE_STATUS: Record<AdminTaskStatus, "queued" | "running" | "done" | "failed" | "cancelled"> = {
  queued: "queued",
  running: "running",
  succeeded: "done",
  failed: "failed",
  cancelled: "cancelled"
};

/** 给租户挂 speaker 槽位（POST /tenants/{id}/voice-slots，幂等：重复挂 changed:false 不报错）。 */
export function assignVoiceSlot(
  tenantId: string,
  input: { speaker_id: string; reason?: string }
): Promise<{ tenant_id: string; speaker_id: string; changed: boolean; speaker_ids: string[] }> {
  return apiFetch(`${BASE}/tenants/${encodeURIComponent(tenantId)}/voice-slots`, { method: "POST", body: input });
}

export function fetchAdminUsage(query: AdminUsageQuery): Promise<AdminUsagePage> {
  return apiFetch<AdminUsagePage>(`${BASE}/usage?${qs({ ...query })}`, { method: "GET" });
}

/**
 * 导出用量 CSV（对账）。CSV 裸响应（BOM 前缀 + Content-Disposition）非信封 → 原始 fetch + 鉴权头；
 * 非 2xx 按信封解析（>50000 行 → 422 USAGE_EXPORT_TOO_LARGE，中文 message 原样展示）。
 */
export async function exportAdminUsageCsv(query: Omit<AdminUsageQuery, "page" | "page_size">): Promise<Blob> {
  const res = await fetch(apiUrl(`${BASE}/usage/export?${qs({ ...query })}`), {
    method: "GET",
    headers: { Accept: "text/csv", ...authHeaders() }
  });
  if (!res.ok) {
    const payload = (await res.json().catch(() => null)) as { error?: { code?: string; message?: string } } | null;
    throw new ApiError(payload?.error?.message ?? "导出失败", payload?.error?.code ?? "EXPORT_FAILED", res.status);
  }
  return res.blob();
}

export function fetchAdminTasks(query: {
  task_family?: AdminTaskFamily | "";
  tenant_id?: string;
  /** BE Literal 含 done 别名（服务端归一 done→succeeded）；UI 只发真枚举。 */
  status?: AdminTaskStatus | "";
  from?: string;
  to?: string;
  page: number;
  page_size: number;
}): Promise<AdminTaskPage> {
  return apiFetch<AdminTaskPage>(`${BASE}/tasks?${qs({ ...query })}`, { method: "GET" });
}

/** 重跑（POST /tasks/{id}/retry?task_family=，202）。仅 retryable；非法 → 409 TASK_NOT_RETRYABLE / 404 TASK_NOT_FOUND / 503 TASK_RETRY_ENQUEUE_FAILED。 */
export function retryAdminTask(taskId: string, taskFamily?: AdminTaskFamily): Promise<AdminRetryReceipt> {
  const suffix = taskFamily ? `?${qs({ task_family: taskFamily })}` : "";
  return apiFetch<AdminRetryReceipt>(`${BASE}/tasks/${encodeURIComponent(taskId)}/retry${suffix}`, { method: "POST" });
}

/** 审计日志（GET /audit-logs——query 仅 action / target_tenant_id / 分页）。只读。 */
export function fetchAdminAudit(query: {
  action?: AuditAction | "";
  target_tenant_id?: string;
  page: number;
  page_size: number;
}): Promise<AdminAuditPage> {
  return apiFetch<AdminAuditPage>(`${BASE}/audit-logs?${qs({ ...query })}`, { method: "GET" });
}

File: backend/app/api/v1/routes/admin_console.py
from __future__ import annotations

import csv
import io
from datetime import date
from typing import Annotated, Literal

from fastapi import APIRouter, Depends, Query, Request
from fastapi import status as http_status
from fastapi.responses import Response
from sqlalchemy.orm import Session

from app.api.deps import CurrentUserDependency, DbSessionDependency, require_platform_admin
from app.core.exceptions import AppError
from app.core.logging import get_logger
from app.db.models import User, VideoTask
from app.schemas.admin_console import (
    AdminAuditLogPage,
    AdminCreditsAdjustRequest,
    AdminCreditsAdjustResponse,
    AdminPlanChangeRequest,
    AdminPlanChangeResponse,
    AdminTaskFamily,
    AdminTaskPage,
    AdminTaskRetryResponse,
    AdminTenantDetail,
    AdminTenantPage,
    AdminTenantStatusRequest,
    AdminTenantStatusResponse,
    AdminUsagePage,
    AdminVoiceSlotAssignRequest,
    AdminVoiceSlotAssignResponse,
    AdminVoiceSlotsResponse,
)
from app.schemas.response import ApiResponse, ok
from app.services import admin_console
from app.services.voice_slots import SpeakerSlotAssignmentError
from app.workers.avatar_talk import generate_avatar_talk_task, generate_seedance_i2v_task
from app.workers.image_gen import generate_ecom_replicate_task, generate_image_task
from app.workers.reverse_prompt import generate_reverse_prompt_video_task
from app.workers.video_gen import generate_video_gen_task
from app.workers.video_tasks import generate_video_task

router = APIRouter(dependencies=[Depends(require_platform_admin)])
logger = get_logger(__name__)

PageQuery = Annotated[int, Query(ge=1)]
PageSizeQuery = Annotated[int, Query(ge=1, le=100)]
FromDateQuery = Annotated[date | None, Query(alias="from")]
ToDateQuery = Annotated[date | None, Query()]


@router.get("/tenants", response_model=ApiResponse[AdminTenantPage])
def list_tenants(
    request: Request,
    q: str | None = Query(default=None, max_length=200),
    plan: Literal["free", "basic", "huading"] | None = None,
    status: Literal["active", "suspended", "closed"] | None = None,
    sort: Literal["credits_used", "created_at", "balance"] = "created_at",
    order: Literal["asc", "desc"] = "desc",
    page: PageQuery = 1,
    page_size: PageSizeQuery = 20,
    db: Session = DbSessionDependency,
) -> ApiResponse[AdminTenantPage]:
    return ok(
        request,
        admin_console.list_tenants(
            db,
            q=q,
            plan=plan,
            status=status,
            sort=sort,
            order=order,
            page=page,
            page_size=page_size,
        ),
    )


@router.get("/tenants/{tenant_id}", response_model=ApiResponse[AdminTenantDetail])
def get_tenant(
    request: Request,
    tenant_id: str,
    db: Session = DbSessionDependency,
) -> ApiResponse[AdminTenantDetail]:
    tenant = admin_console.tenant_item_or_404(db, tenant_id=tenant_id)
    slots = [item for item in admin_console.voice_slots(db).items if item.tenant_id == tenant_id]
    return ok(
        request,
        AdminTenantDetail(
            tenant=tenant,
            recent_tasks=admin_console.recent_tasks(db, tenant_id=tenant_id),
            recent_usage=admin_console.recent_usage(db, tenant_id=tenant_id),
            voice_slots=slots,
        ),
    )


@router.post(
    "/tenants/{tenant_id}/credits",
    response_model=ApiResponse[AdminCreditsAdjustResponse],
)
def adjust_tenant_credits(
    request: Request,
    tenant_id: str,
    payload: AdminCreditsAdjustRequest,
    user: User = CurrentUserDependency,
    db: Session = DbSessionDependency,
) -> ApiResponse[AdminCreditsAdjustResponse]:
    try:
        result = admin_console.adjust_credits(
            db,
            actor=user,
            tenant_id=tenant_id,
            delta=payload.delta,
            reason=payload.reason,
        )
        db.commit()
    except Exception:
        db.rollback()
        raise
    return ok(request, result)


@router.patch(
    "/tenants/{tenant_id}/plan",
    response_model=ApiResponse[AdminPlanChangeResponse],
)
def change_tenant_plan(
    request: Request,
    tenant_id: str,
    payload: AdminPlanChangeRequest,
    user: User = CurrentUserDependency,
    db: Session = DbSessionDependency,
) -> ApiResponse[AdminPlanChangeResponse]:
    try:
        result = admin_console.change_plan(
            db,
            actor=user,
            tenant_id=tenant_id,
            plan_code=payload.plan_code,
            reason=payload.reason,
        )
        db.commit()
    except Exception:
        db.rollback()
        raise
    return ok(request, result)


@router.patch(
    "/tenants/{tenant_id}/status",
    response_model=ApiResponse[AdminTenantStatusResponse],
)
def change_tenant_status(
    request: Request,
    tenant_id: str,
    payload: AdminTenantStatusRequest,
    user: User = CurrentUserDependency,
    db: Session = DbSessionDependency,
) -> ApiResponse[AdminTenantStatusResponse]:
    try:
        result = admin_console.change_tenant_status(
            db,
            actor=user,
            tenant_id=tenant_id,
            active=payload.active,
            reason=payload.reason,
        )
        db.commit()
    except Exception:
        db.rollback()
        raise
    return ok(request, result)


@router.get("/voice-slots", response_model=ApiResponse[AdminVoiceSlotsResponse])
def get_voice_slots(
    request: Request,
    db: Session = DbSessionDependency,
) -> ApiResponse[AdminVoiceSlotsResponse]:
    return ok(request, admin_console.voice_slots(db))


@router.post(
    "/tenants/{tenant_id}/voice-slots",
    response_model=ApiResponse[AdminVoiceSlotAssignResponse],
)
def assign_voice_slot(
    request: Request,
    tenant_id: str,
    payload: AdminVoiceSlotAssignRequest,
    user: User = CurrentUserDependency,
    db: Session = DbSessionDependency,
) -> ApiResponse[AdminVoiceSlotAssignResponse]:
    try:
        result = admin_console.assign_tenant_voice_slot(
            db,
            actor=user,
            tenant_id=tenant_id,
            speaker_id=payload.speaker_id,
            reason=payload.reason,
        )
        db.commit()
    except SpeakerSlotAssignmentError as exc:
        db.rollback()
        raise AppError(
            str(exc),
            code="VOICE_SLOT_ASSIGNMENT_FAILED",
            status_code=422,
        ) from exc
    except Exception:
        db.rollback()
        raise
    return ok(request, result)


@router.get("/usage", response_model=ApiResponse[AdminUsagePage])
def get_usage(
    request: Request,
    tenant_id: str | None = None,
    from_: FromDateQuery = None,
    to: ToDateQuery = None,
    capability: str | None = Query(default=None, max_length=32),
    provider: str | None = Query(default=None, max_length=40),
    status: Literal["reserved", "settled", "released"] | None = None,
    page: PageQuery = 1,
    page_size: PageSizeQuery = 20,
    db: Session = DbSessionDependency,
) -> ApiResponse[AdminUsagePage]:
    return ok(
        request,
        admin_console.list_usage(
            db,
            tenant_id=tenant_id,
            from_=from_,
            to=to,
            capability=capability,
            provider=provider,
            status=status,
            page=page,
            page_size=page_size,
        ),
    )


@router.get("/usage/export", response_class=Response)
def export_usage(
    tenant_id: str | None = None,
    from_: FromDateQuery = None,
    to: ToDateQuery = None,
    capability: str | None = Query(default=None, max_length=32),
    provider: str | None = Query(default=None, max_length=40),
    status: Literal["reserved", "settled", "released"] | None = None,
    db: Session = DbSessionDependency,
) -> Response:
    rows = admin_console.export_usage_rows(
        db,
        tenant_id=tenant_id,
        from_=from_,
        to=to,
        capability=capability,
        provider=provider,
        status=status,
    )
    output = io.StringIO(newline="")
    writer = csv.writer(output)
    writer.writerow(
        [
            "created_at",
            "tenant_id",
            "tenant_slug",
            "tenant_name",
            "capability",
            "provider",
            "model",
            "quantity",
            "unit",
            "credits",
            "cost_cents",
            "status",
            "video_task_id",
        ]
    )
    for item in rows:
        writer.writerow(
            [
                item.created_at.isoformat(),
                item.tenant_id,
                item.tenant_slug,
                item.tenant_name,
                item.capability,
                item.provider,
                item.model or "",
                item.quantity,
                item.unit,
                item.credits,
                item.cost_cents,
                item.status,
                item.video_task_id or "",
            ]
        )
    return Response(
        content="\ufeff" + output.getvalue(),
        media_type="text/csv",
        headers={"Content-Disposition": 'attachment; filename="usage-records.csv"'},
    )


@router.get("/tasks", response_model=ApiResponse[AdminTaskPage])
def get_tasks(
    request: Request,
    task_family: AdminTaskFamily | None = None,
    tenant_id: str | None = None,
    status: Literal["queued", "running", "done", "succeeded", "failed", "cancelled"]
    | None = None,
    from_: FromDateQuery = None,
    to: ToDateQuery = None,
    page: PageQuery = 1,
    page_size: PageSizeQuery = 20,
    db: Session = DbSessionDependency,
) -> ApiResponse[AdminTaskPage]:
    return ok(
        request,
        admin_console.list_tasks(
            db,
            task_family=task_family,
            tenant_id=tenant_id,
            status=status,
            from_=from_,
            to=to,
            page=page,
            page_size=page_size,
        ),
    )


def _enqueue_task_retry(task: VideoTask) -> None:
    payload = admin_console.task_retry_payload(task)
    if task.video_mode == "photo":
        generate_image_task.apply_async(args=[payload], task_id=task.id, queue="image")
    elif task.video_mode == "seedance_i2v":
        generate_seedance_i2v_task.apply_async(args=[payload], task_id=task.id, queue="video")
    elif task.video_mode == "video_gen":
        generate_video_gen_task.apply_async(args=[payload], task_id=task.id, queue="video")
    elif task.video_mode == "avatar_talk":
        generate_avatar_talk_task.apply_async(args=[payload], task_id=task.id, queue="avatar")
    else:
        generate_video_task.apply_async(args=[payload], task_id=task.id, queue="default")


def _compensate_retry_dispatch_failure(
    db: Session,
    *,
    actor_id: str,
    task_family: AdminTaskFamily,
    task_id: str,
    output_indexes: list[int],
) -> None:
    try:
        with Session(bind=db.get_bind()) as compensation_db:
            actor = compensation_db.get(User, actor_id)
            if actor is None:
                raise RuntimeError("Retry compensation actor was not found.")
            if task_family == "video":
                admin_console.compensate_task_retry_enqueue_failure(
                    compensation_db,
                    actor=actor,
                    task_id=task_id,
                )
            elif task_family == "reverse_prompt":
                admin_console.compensate_reverse_prompt_retry_enqueue_failure(
                    compensation_db,
                    actor=actor,
                    job_id=task_id,
                )
            else:
                admin_console.compensate_ecom_replicate_retry_enqueue_failure(
                    compensation_db,
                    actor=actor,
                    job_id=task_id,
                    output_indexes=output_indexes,
                )
            compensation_db.commit()
    except Exception:
        logger.exception(
            "admin_task_retry_compensation_failed",
            task_family=task_family,
            task_id=task_id,
        )


@router.post(
    "/tasks/{task_id}/retry",
    response_model=ApiResponse[AdminTaskRetryResponse],
    status_code=http_status.HTTP_202_ACCEPTED,
)
def retry_task(
    request: Request,
    task_id: str,
    task_family: AdminTaskFamily | None = None,
    user: User = CurrentUserDependency,
    db: Session = DbSessionDependency,
) -> ApiResponse[AdminTaskRetryResponse]:
    try:
        resolved_family = admin_console.resolve_task_family(
            db,
            task_id=task_id,
            requested_family=task_family,
        )
        if resolved_family == "video":
            preparation = admin_console.prepare_task_retry(db, actor=user, task_id=task_id)
        elif resolved_family == "reverse_prompt":
            preparation = admin_console.prepare_reverse_prompt_retry(
                db,
                actor=user,
                job_id=task_id,
            )
        else:
            preparation = admin_console.prepare_ecom_replicate_retry(
                db,
                actor=user,
                job_id=task_id,
            )
        task = preparation.task
        retried_output_indexes = list(preparation.output_indexes)
        response = AdminTaskRetryResponse(
            id=task.id,
            task_family=resolved_family,
            tenant_id=task.tenant_id,
            status="queued",
            progress=task.progress if resolved_family == "video" else 0,
            charged=preparation.charged,
            credits=preparation.credits,
            is_estimate=preparation.is_estimate,
        )
        db.commit()
    except Exception:
        db.rollback()
        raise
    try:
        if resolved_family == "video":
            _enqueue_task_retry(task)
        elif resolved_family == "reverse_prompt":
            generate_reverse_prompt_video_task.apply_async(
                args=[task.id], task_id=task.id, queue="image"
            )
        else:
            generate_ecom_replicate_task.apply_async(
                args=[task.id], task_id=task.id, queue="image"
            )
    except Exception as exc:
        _compensate_retry_dispatch_failure(
            db,
            actor_id=user.id,
            task_family=resolved_family,
            task_id=task.id,
            output_indexes=retried_output_indexes,
        )
        raise AppError(
            "Task retry could not be queued. Please retry.",
            code="TASK_RETRY_ENQUEUE_FAILED",
            status_code=503,
        ) from exc
    return ok(request, response)


@router.get("/audit-logs", response_model=ApiResponse[AdminAuditLogPage])
def get_audit_logs(
    request: Request,
    action: Literal[
        "credits_adjust",
        "plan_change",
        "status_change",
        "voice_slot_assign",
        "task_retry",
    ]
    | None = None,
    target_tenant_id: str | None = None,
    page: PageQuery = 1,
    page_size: PageSizeQuery = 20,
    db: Session = DbSessionDependency,
) -> ApiResponse[AdminAuditLogPage]:
    return ok(
        request,
        admin_console.list_audit_logs(
            db,
            action=action,
            target_tenant_id=target_tenant_id,
            page=page,
            page_size=page_size,
        ),
    )
